IPIP/0000-gateway-json-cbor-response-format.md

```diff
@@ -0,0 +1,102 @@
+# IPIP 0000: Gateway JSON and TAR Response Formats
+
+- Start Date: 2022-10-07
+- Related Issues:
+  - [ipfs/in-web-browsers/issues/182]
+  - [ipfs/specs/pull/328]
+  - [ipfs/kubo/issues/8823]
+  - [ipfs/kubo/pull/9335]
+  - [ipfs/go-ipfs/issues/7552]
+
+## Summary
+
+Add [DAG-JSON], [DAG-CBOR], JSON and CBOR response formats to the [HTTP Gateway](../http-gateways/).
+
+## Motivation
+
+Currently, the gateway supports requesting data in the [DAG-PB], RAW, and [CAR]
+formats. However, it should be possible to download deserialized  version of  data other than UnixFS
+in order to unlock the potential of the [IPLD Data Model][ipld-data-model] beyond files and directories.
+
+
+The main functional gap in IPFS ecosystem is the lack of support for non-UnixFS DAGs on HTTP Gateways.
+Users are able to create custom DAGs based on traversable DAG-CBOR thanks to [CBOR tag 42 being reserved for CIDs](https://github.com/core-wg/yang-cbor/issues/13#issuecomment-524378859), 
+but they are unable to load deserialized  DAG-CBOR documents from local gateway, 
+which is severely decreasing the utility of non-UnixFS DAGs.
+
+Adding new responses types will also benefit UnixFS.
+DAG-PB has a [logical format][dag-pb-format] which makes it possible
+to represent a DAG-PB directory as a [DAG-JSON] document. This means that, if we
+support DAG-JSON in the gateway, then we would support
+[JSON responses for directory listings][ipfs/go-ipfs/issues/7552], which has been
+requested by our users in the past.
+
+In addition, this functionality is already present on the current Kubo CLI. By
+bringing it to the gateways, we provide users with more power when it comes
+to traversing IPLD data.
+
+## Detailed design
+
+The solution is to allow the Gateway to support serializing an IPLD Data Model
+representation as [DAG-JSON], [DAG-CBOR], JSON and CBOR by requesting them
+using either the `Accept` HTTP header or the `format` URL query.
+
+## Test fixtures
+
+- [`bafybeiegxwlgmoh2cny7qlolykdf7aq7g6dlommarldrbm7c4hbckhfcke`][dag-pb] is a
+  DAG-PB directory.
+- [`bafkreidmwhhm6myajxlpu7kofe3aqwf4ezxxn46cp5fko7mb6x74g4k5nm`][dag-pb-json]
+  is the aforementioned DAG-PB directory's Logical DAG-JSON representation that
+  is expected to be returned when using `?format=dag-json`.
+- Traversal Test Fixtures: the following test fixtures contain two levels of nested
+  documents of their encoding. Accessing `/ipfs/$CID/foo/bar` should return the JSON
+  equivalent of `{"hello":"this is not a link"}`.
+  - DAG-CBOR: [`bafyreiehxu373cu3v5gyxyxfsfjryscs7sq6fh3unqcqgqhdfn3n43vrgu`][dag-cbor-traversal]
+  - DAG-JSON: [`baguqeeraoaeabj5hdfcmpkzfeiwtfwb3qbvfwzbiknqn7itcwsb2fdtu7eta`][dag-json-traversal]
+- `TODO` is a valid JSON but not a valid DAG-JSON
+- `TODO` is a valid CBOR but not a valid DAG-CBOR
```

Comment:
@lidel Unfortunately, I'm not sure if this will be easy to do. See https://github.com/ipfs/kubo/pull/9335#issuecomment-1277057863. It seems that the Go codecs encoders are quite strict, making the differences between both codecs very minimal.